File: whdon_app/src/main/java/nl/whitedove/washetdroogofniet/WeerHelper.java
package nl.whitedove.washetdroogofniet;

import android.annotation.SuppressLint;
import android.content.Context;
import android.net.Uri;
import android.support.v4.content.ContextCompat;

import org.joda.time.DateTime;
import org.joda.time.Minutes;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.Random;
import java.util.regex.Pattern;

import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

class WeerHelper {

    public enum WeerType {
        Onbekend(0L), Zonnig(1L), Halfbewolkt(2L), Bewolkt(3L), Regen(9L), Buien(10L), Onweer(11L), Sneeuw(13L), Mist(50L);

        private long value;
        @SuppressLint("UseSparseArrays")
        private static Map<Long, WeerType> map = new HashMap<>();

        WeerType(long value) {
            this.value = value;
        }

        static {
            for (WeerType weerType : WeerType.values()) {
                map.put(weerType.value, weerType);
            }
        }

        public static WeerType valueOf(long weerType) {
            return map.get(weerType);
        }

        public long getValue() {
            return value;
        }
    }

    public enum WindDirection {
        Onbekend(0L), Noord(1L), NoordOost(2L), Oost(3L), ZuidOost(4L), Zuid(5L), ZuidWest(6L), West(7L), NoordWest(8L);

        private long value;
        @SuppressLint("UseSparseArrays")
        private static Map<Long, WindDirection> map = new HashMap<>();

        WindDirection(long value) {
            this.value = value;
        }

        static {
            for (WindDirection windDirection : WindDirection.values()) {
                map.put(windDirection.value, windDirection);
            }
        }

        public static WindDirection valueOf(long windDirection) {
            return map.get(windDirection);
        }

        public long getValue() {
            return value;
        }
    }

    private static int HuidigeTemperatuur = 999;

    private static WeerType HuidigeWeertype = WeerType.Onbekend;

    private static int HuidigeWindSpeed = 999;

    private static WindDirection HuidigeWindDir = WindDirection.Onbekend;

    static Weer BepaalWeer() throws JSONException {

        String weatherJson = WeerHelper.getWeatherData();
        if (weatherJson == null || weatherJson.isEmpty()) {
            return null;
        }

        JSONObject jObj = new JSONObject(weatherJson);
        JSONArray jArrWeather = jObj.getJSONArray("weather");
        JSONObject main = jObj.getJSONObject("main");
        JSONObject weather = jArrWeather.getJSONObject(0);
        JSONObject wind = jObj.getJSONObject("wind");

        Weer result = new Weer();
        String plaats = jObj.getString("name");
        plaats = plaats.replace("Gemeente", "");
        result.setPlaats(plaats);
        if (weather.has("icon")) result.setIcon(weather.getString("icon"));
        if (main.has("temp")) result.setGraden((int) Math.round(main.getDouble("temp")));
        result.setWind(wind.has("speed") ? (int) Math.round(3.6f * wind.getDouble("speed")) : 0);
        int windRichting = wind.has("deg") ? (int) Math.round(wind.getDouble("deg")) : -1;
        result.setWindDir(windRichting);
        return result;
    }

    static String BepaalBrDataTxt(Context cxt, BuienData buienData) {
        if (buienData == null || buienData.getNoData()) return cxt.getString(R.string.NoBrData);
        return cxt.getString(R.string.DroogTxt);
    }

    static BuienData BepaalBuien() throws IOException {

        BuienData result = new BuienData();
        String brString = null;

        if (Helper.mCurrentBestLocation != null) {

            Double lat = Helper.mCurrentBestLocation.getLatitude();
            Double lon = Helper.mCurrentBestLocation.getLongitude();

            String sLat = String.format(Locale.ROOT, "%.6f", lat);
            String sLon = String.format(Locale.ROOT, "%.6f", lon);

            brString = WeerHelper.getBrData(sLat, sLon);
        }

        if (brString == null || brString.isEmpty()) {
            result.setNoData(true);
            ArrayList<RegenEntry> brData = new ArrayList<>();
            // Add some fake data
            for (int i = 0; i < 10; i++) {
                brData.add(new RegenEntry("", 0));
            }
            result.setRegenData(brData);
            return result;
        }

        result.setNoData(false);

        int i = 0;
        ArrayList<RegenEntry> regenData = new ArrayList<>();
        String[] parts = brString.split("\r\n");
        for (String entry : parts) {

            String[] tr = entry.split(Pattern.quote("|"));
            RegenEntry re = new RegenEntry();

            if (Helper.DEBUG) {
                Random rnd = new Random();
                i += rnd.nextInt(20) - 10;
            } else {
                i = 0;
            }

            if (Helper.tryParseInt(tr[0]))
                re.setRegen(i + Integer.parseInt(tr[0]));
            else
                re.setRegen(i);
            re.setTijd(tr[1]);
            regenData.add(re);
        }

        result.setRegenData(regenData);
        return result;
    }

    private static String getBrData(String lat, String lon) {

        Uri.Builder builder = new Uri.Builder();
        builder.scheme("http")
                .authority("gadgets.buienradar.nl")
                .appendPath("data")
                .appendPath("raintext")
                .appendQueryParameter("lat", lat)
                .appendQueryParameter("lon", lon);

        String url = builder.build().toString();
        Helper.Log("Buienradar url:" + url);

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .build();

        String brData = null;
        Response response;
        try {
            response = client.newCall(request).execute();
            if (response.isSuccessful()) //noinspection ConstantConditions
                brData = response.body().string();
        } catch (IOException ignored) {
        }
        return brData;
    }

    private static String getWeatherData() {

        String locatie = LocationHelper.GetLocatieVoorWeer();
        String country = LocationHelper.GetCountryVoorWeer();

        Uri.Builder builder = new Uri.Builder();
        builder.scheme("http")
                .authority("api.openweathermap.org")
                .appendPath("data")
                .appendPath("2.5")
                .appendPath("weather")
                .appendQueryParameter("q", String.format("%s,%s", locatie, country))
                .appendQueryParameter("appid", "e246fa149696b433128c8e774219bbc8")
                .appendQueryParameter("cnt", "1")
                .appendQueryParameter("units", "metric");

        String url = builder.build().toString();
        Helper.Log("weather url:" + url);

        OkHttpClient client = new OkHttpClient();
        Request request = new Request.Builder()
                .url(url)
                .build();

        String weatherData = null;
        Response response;
        try {
            response = client.newCall(request).execute();
            if (response.isSuccessful()) //noinspection ConstantConditions
                weatherData = response.body().string();
        } catch (IOException ignored) {
        }
        return weatherData;
    }

    static float BerekenNuXPositie(BuienData buienData) {
        if (buienData == null || buienData.getNoData()) return 0f;
        DateTime nu = DateTime.now();
        RegenEntry eersteEntry = buienData.getRegenData().get(0);
        String tijd = eersteEntry.getTijd();
        String sUur = tijd.substring(0, 2);
        String sMin = tijd.substring(3);
        int iUur = Integer.parseInt(sUur);
        int iMin = Integer.parseInt(sMin);
        DateTime first = new DateTime(nu.getYear(), nu.getMonthOfYear(), nu.getDayOfMonth(), iUur, iMin);
        if (first.isAfter(nu)) first = first.minusDays(1);
        int minuten = Minutes.minutesBetween(first, nu).getMinutes();
        return 24f * (minuten / 120f);
    }

    static int GetHuidigeTemperatuur() {
        return HuidigeTemperatuur;
    }

    static void SetHuidigeTemperatuur(int temperatuur) {
        HuidigeTemperatuur = temperatuur;
    }

    static WeerType getHuidigeWeertype() {
        return HuidigeWeertype;
    }

    static void setHuidigeWeertype(WeerType huidigeWeertype) {
        HuidigeWeertype = huidigeWeertype;
    }

    static int getHuidigeWindSpeed() {
        return HuidigeWindSpeed;
    }

    static void setHuidigeWindSpeed(int huidigeWindSpeed) {
        HuidigeWindSpeed = huidigeWindSpeed;
    }

    static WindDirection getHuidigeWindDir() {
        return HuidigeWindDir;
    }

    static void setHuidigeWindDir(WindDirection huidigeWindDir) {
        HuidigeWindDir = huidigeWindDir;
    }

    @SuppressLint("DefaultLocale")
    static String WeerTypeToWeerIcoon(WeerType weerType) {
        if (weerType == WeerType.Onbekend) {
            return null;
        } else {
            return String.format("i%02dd", weerType.getValue());
        }
    }

    static WeerType WeerIcoonToWeerType(String weerIcoon) {
        switch (weerIcoon) {
            case "01d":
            case "01n":
                return WeerType.Zonnig;
            case "02d":
            case "02n":
                return WeerType.Halfbewolkt;
            case "03d":
            case "03n":
            case "04d":
            case "04n":
                return WeerType.Bewolkt;
            case "09d":
            case "09n":
                return WeerType.Regen;
            case "10d":
            case "10n":
                return WeerType.Buien;
            case "11d":
            case "11n":
                return WeerType.Onweer;
            case "13d":
            case "13n":
                return WeerType.Sneeuw;
            case "50d":
            case "50n":
                return WeerType.Mist;
        }
        return WeerType.Onbekend;
    }

    static String WeerTypeToWeerOmschrijving(WeerType weerType) {
        switch (weerType) {
            case Onbekend:
                return null;
            case Zonnig:
                return "Zonnig";
            case Halfbewolkt:
                return "Halfbewolkt";
            case Bewolkt:
                return "Bewolkt";
            case Buien:
                return "Buien";
            case Mist:
                return "Mist";
            case Onweer:
                return "Onweer";
            case Regen:
                return "Regen";
            case Sneeuw:
                return "Sneeuw";
        }
        return null;
    }

    static Integer WeerTypeToWeerKleur(Context context, WeerType weerType) {
        switch (weerType) {
            case Onbekend:
                return ContextCompat.getColor(context, R.color.colorPrimary);
            case Zonnig:
                return ContextCompat.getColor(context, R.color.colorGrafiek5);
            case Halfbewolkt:
                return ContextCompat.getColor(context, R.color.colorGrafiek4);
            case Bewolkt:
                return ContextCompat.getColor(context, R.color.colorGrafiek3);
            case Buien:
                return ContextCompat.getColor(context, R.color.colorGrafiek6);
            case Mist:
                return ContextCompat.getColor(context, R.color.colorGrafiek7);
            case Onweer:
                return ContextCompat.getColor(context, R.color.colorGrafiek2);
            case Regen:
                return ContextCompat.getColor(context, R.color.colorGrafiek1);
            case Sneeuw:
                return ContextCompat.getColor(context, R.color.colorGrafiek8);
        }
        return null;
    }

    static String WindDirectionToOmschrijving(WindDirection windDir) {
        switch (windDir) {
            case Onbekend:
                return null;
            case Noord:
                return "Noord";
            case NoordOost:
                return "Noordoost";
            case Oost:
                return "Oost";
            case ZuidOost:
                return "Zuidoost";
            case Zuid:
                return "Zuid";
            case ZuidWest:
                return "Zuidwest";
            case West:
                return "West";
            case NoordWest:
                return "Noordwest";
        }
        return null;
    }

}

File: whdon_app/src/main/java/nl/whitedove/washetdroogofniet/WeerHelper.kt
package nl.whitedove.washetdroogofniet

import android.annotation.SuppressLint
import android.content.Context
import android.net.Uri
import android.support.v4.content.ContextCompat
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import org.joda.time.DateTime
import org.joda.time.Minutes
import org.json.JSONException
import org.json.JSONObject
import java.io.IOException
import java.util.*
import java.util.regex.Pattern

object WeerHelper {

    private var HuidigeTemperatuur = 999
    var huidigeWeertype = WeerType.Onbekend
    var huidigeWindSpeed = 999
    var huidigeWindDir = WindDirection.Onbekend

    enum class WeerType(val value: Long) {
        Onbekend(0L), Zonnig(1L), Halfbewolkt(2L), Bewolkt(3L), Regen(9L), Buien(10L), Onweer(11L), Sneeuw(13L), Mist(50L);


        companion object {
            @SuppressLint("UseSparseArrays")
            private val map = HashMap<Long, WeerType>()

            init {
                for (weerType in WeerType.values()) {
                    map[weerType.value] = weerType
                }
            }

            fun valueOf(weerType: Long): WeerType {
                return map[weerType] ?: Onbekend
            }
        }
    }

    enum class WindDirection(val value: Long) {
        Onbekend(0L), Noord(1L), NoordOost(2L), Oost(3L), ZuidOost(4L), Zuid(5L), ZuidWest(6L), West(7L), NoordWest(8L);

        companion object {
            @SuppressLint("UseSparseArrays")
            private val map = HashMap<Long, WindDirection>()

            init {
                for (windDirection in WindDirection.values()) {
                    map[windDirection.value] = windDirection
                }
            }

            fun valueOf(windDirection: Long): WindDirection {
                return map[windDirection] ?: Onbekend
            }
        }
    }

    private fun getweatherData(): String? {

        if (Helper.mCurrentBestLocation == null) return null

        val lat = Helper.mCurrentBestLocation!!.latitude
        val lon = Helper.mCurrentBestLocation!!.longitude
        val sLat = String.format(Locale.ROOT, "%.6f", lat)
        val sLon = String.format(Locale.ROOT, "%.6f", lon)

        val builder = Uri.Builder()
        builder.scheme("http")
                .authority("api.openweathermap.org")
                .appendPath("data")
                .appendPath("2.5")
                .appendPath("weather")
                .appendQueryParameter("lat", sLat)
                .appendQueryParameter("lon", sLon)
                .appendQueryParameter("appid", "e246fa149696b433128c8e774219bbc8")
                .appendQueryParameter("cnt", "1")
                .appendQueryParameter("units", "metric")

        val url = builder.build().toString()
        Helper.l("weather url:$url")

        val client = OkHttpClient()
        val request = Request.Builder()
                .addHeader("Cache-Control", "no-cache")
                .url(url)
                .build()

        var weatherData: String? = null
        val response: Response
        try {
            response = client.newCall(request).execute()
            if (response.isSuccessful)
                weatherData = response.body()!!.string()
        } catch (ignored: IOException) {
        }

        return weatherData
    }

    @Throws(JSONException::class)
    fun bepaalWeer(): Weer? {

        val weatherJson = WeerHelper.getweatherData()
        if (weatherJson == null || weatherJson.isEmpty()) {
            return null
        }

        val jObj = JSONObject(weatherJson)
        val jArrWeather = jObj.getJSONArray("weather")
        val main = jObj.getJSONObject("main")
        val weather = jArrWeather.getJSONObject(0)
        val wind = jObj.getJSONObject("wind")

        val result = Weer()
        var plaats = jObj.getString("name")
        plaats = plaats.replace("Gemeente", "")
        result.plaats = plaats
        if (weather.has("icon")) result.icon = weather.getString("icon")
        if (main.has("temp")) result.graden = Math.round(main.getDouble("temp")).toInt()
        result.wind = if (wind.has("speed")) Math.round(3.6f * wind.getDouble("speed")).toInt() else 0
        val windRichting = if (wind.has("deg")) Math.round(wind.getDouble("deg")).toInt() else -1
        result.setWindDir(windRichting)
        return result
    }

    fun bepaalBrDataTxt(cxt: Context, buienData: BuienData?): String {
        return if (buienData == null || buienData.noData!!) cxt.getString(R.string.NoBrData) else cxt.getString(R.string.DroogTxt)
    }

    fun bepaalBuien(): BuienData {

        val result = BuienData()
        var brString: String? = null

        if (Helper.mCurrentBestLocation != null) {

            val lat = Helper.mCurrentBestLocation!!.latitude
            val lon = Helper.mCurrentBestLocation!!.longitude
            val sLat = String.format(Locale.ROOT, "%.6f", lat)
            val sLon = String.format(Locale.ROOT, "%.6f", lon)

            brString = WeerHelper.getBrData(sLat, sLon)
        }

        if (brString == null || brString.isEmpty()) {
            result.noData = true
            val brData = ArrayList<RegenEntry>()
            // Add some fake data
            for (i in 0..9) {
                brData.add(RegenEntry("", 0))
            }
            result.regenData = brData
            return result
        }

        result.noData = false

        val regenData = ArrayList<RegenEntry>()
        val parts = brString.split("\r\n".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        for (entry in parts) {

            val tr = entry.split(Pattern.quote("|").toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
            val re = RegenEntry()

            if (Helper.tryParseInt(tr[0]))
                re.regen = Integer.parseInt(tr[0])
            else
                re.regen = 0
            re.tijd = tr[1]
            regenData.add(re)
        }

        result.regenData = regenData
        return result
    }

    private fun getBrData(lat: String, lon: String): String? {

        val builder = Uri.Builder()
        builder.scheme("http")
                .authority("gadgets.buienradar.nl")
                .appendPath("data")
                .appendPath("raintext")
                .appendQueryParameter("lat", lat)
                .appendQueryParameter("lon", lon)

        val url = builder.build().toString()
        Helper.l("Buienradar url:$url")

        val client = OkHttpClient()
        val request = Request.Builder()
                .addHeader("Cache-Control", "no-cache")
                .url(url)
                .build()

        var brData: String? = null
        val response: Response
        try {
            response = client.newCall(request).execute()
            if (response.isSuccessful)

                brData = response.body()!!.string()
        } catch (ignored: IOException) {
        }

        return brData
    }

    fun berekenNuXPositie(buienData: BuienData?): Float {
        if (buienData == null || buienData.noData!!) return 0f
        val nu = DateTime.now()
        val eersteEntry = buienData.regenData!![0]
        val tijd = eersteEntry.tijd
        val sUur = tijd.substring(0, 2)
        val sMin = tijd.substring(3)
        val iUur = Integer.parseInt(sUur)
        val iMin = Integer.parseInt(sMin)
        var first = DateTime(nu.year, nu.monthOfYear, nu.dayOfMonth, iUur, iMin)
        if (first.isAfter(nu)) first = first.minusDays(1)
        val minuten = Minutes.minutesBetween(first, nu).minutes
        return 24f * (minuten / 120f)
    }

    fun getHuidigeTemperatuur(): Int {
        return HuidigeTemperatuur
    }

    fun setHuidigeTemperatuur(temperatuur: Int) {
        HuidigeTemperatuur = temperatuur
    }

    @SuppressLint("DefaultLocale")
    fun weerTypeToWeerIcoon(weerType: WeerType): String? {
        return if (weerType == WeerType.Onbekend) {
            null
        } else {
            String.format("i%02dd", weerType.value)
        }
    }

    fun weerIcoonToWeerType(weerIcoon: String): WeerType {
        when (weerIcoon) {
            "01d", "01n" -> return WeerType.Zonnig
            "02d", "02n" -> return WeerType.Halfbewolkt
            "03d", "03n", "04d", "04n" -> return WeerType.Bewolkt
            "09d", "09n" -> return WeerType.Regen
            "10d", "10n" -> return WeerType.Buien
            "11d", "11n" -> return WeerType.Onweer
            "13d", "13n" -> return WeerType.Sneeuw
            "50d", "50n" -> return WeerType.Mist
        }
        return WeerType.Onbekend
    }

    fun weerTypeToWeerOmschrijving(weerType: WeerType): String {
        return when (weerType) {
            WeerHelper.WeerType.Onbekend -> ""
            WeerHelper.WeerType.Zonnig -> "Zonnig"
            WeerHelper
                    .WeerType.Halfbewolkt -> "Halfbewolkt"
            WeerHelper.WeerType.Bewolkt -> "Bewolkt"
            WeerHelper.WeerType.Buien -> "Buien"
            WeerHelper.WeerType.Mist -> "Mist"
            WeerHelper.WeerType.Onweer -> "Onweer"
            WeerHelper.WeerType.Regen -> "Regen"
            WeerHelper.WeerType.Sneeuw -> "Sneeuw"
        }
    }

    fun weerTypeToWeerKleur(context: Context, weerType: WeerType): Int {
        return when (weerType) {
            WeerHelper.WeerType.Onbekend -> ContextCompat.getColor(context, R.color.colorPrimary)
            WeerHelper.WeerType.Zonnig -> ContextCompat.getColor(context, R.color.colorGrafiek5)
            WeerHelper.WeerType.Halfbewolkt -> ContextCompat.getColor(context, R.color.colorGrafiek4)
            WeerHelper.WeerType.Bewolkt -> ContextCompat.getColor(context, R.color.colorGrafiek3)
            WeerHelper.WeerType.Buien -> ContextCompat.getColor(context, R.color.colorGrafiek6)
            WeerHelper.WeerType.Mist -> ContextCompat.getColor(context, R.color.colorGrafiek7)
            WeerHelper.WeerType.Onweer -> ContextCompat.getColor(context, R.color.colorGrafiek2)
            WeerHelper.WeerType.Regen -> ContextCompat.getColor(context, R.color.colorGrafiek1)
            WeerHelper.WeerType.Sneeuw -> ContextCompat.getColor(context, R.color.colorGrafiek8)
        }
    }

    fun windDirectionToOmschrijving(windDir: WindDirection): String {
        return when (windDir) {
            WeerHelper.WindDirection.Onbekend -> ""
            WeerHelper.WindDirection.Noord -> "Noord"
            WeerHelper.WindDirection.NoordOost -> "Noordoost"
            WeerHelper.WindDirection.Oost -> "Oost"
            WeerHelper.WindDirection.ZuidOost -> "Zuidoost"
            WeerHelper.WindDirection.Zuid -> "Zuid"
            WeerHelper.WindDirection.ZuidWest -> "Zuidwest"
            WeerHelper.WindDirection.West -> "West"
            WeerHelper.WindDirection.NoordWest -> "Noordwest"
        }
    }
}
